### ISSUE TITLE
Possibly inaccurate assert() in gc.c

### ISSUE THREAD

**Issue by rbehrends** [JuliaLang/julia#38479]
We have observed a possibly broken assert() condition in `src/gc.c` ([link](https://github.com/JuliaLang/julia/blob/c9a149b46196629b6547b906dab444c8a06dc262/src/gc.c#L2665)).

It says:

```c
assert(layout->nfields > 0 && layout->fielddesc_type != 3 && "opaque types should have been handled specially");
```

However, immediately following is a series of tests, with the last branch specifically handling the case of `layout->fielddesc_type == 3`, and that is also where it should normally happen (this is the layout type where GC marking is determined by a custom mark function).

I suspect that the assert() should actually be something like:

```
assert((layout->nfields > 0 || layout->fielddesc_type == 3) && "opaque types should have been handled specially"); 
```

We noticed it, because we hit errors when running with Julia assertions enabled in [GAP.jl](https://github.com/oscar-system/GAP.jl).

This should normally have been caught by the tests in `test/gcext`, but it appears I failed to hook up that particular test to the Julia test harness when submitting the patch for the GC extensions. So the test isn't actually run normally; and as a result, the code in `test/gcext` itself also isn't up to date for 1.6. I plan on submitting a PR that fixes that also.

**Comment by vtjnash** [comment]
SGTM

**Comment by fingolfin** [comment]
I also notice that `fielddesc_type == 3` is not really documented and checked for in several places. @rbehrends I suggest to include something like this in your PR, too, as a safeguard:
```patch
diff --git a/src/julia.h b/src/julia.h
index 7560886301..6730622b82 100644
--- a/src/julia.h
+++ b/src/julia.h
@@ -442,7 +442,7 @@ typedef struct {
     int32_t first_ptr; // index of the first pointer (or -1)
     uint16_t alignment; // strictest alignment over all fields
     uint16_t haspadding : 1; // has internal undefined bytes
-    uint16_t fielddesc_type : 2; // 0 -> 8, 1 -> 16, 2 -> 32
+    uint16_t fielddesc_type : 2; // 0 -> 8, 1 -> 16, 2 -> 32, 3 -> foreign type
     // union {
     //     jl_fielddesc8_t field8[nfields];
     //     jl_fielddesc16_t field16[nfields];
@@ -1030,6 +1030,7 @@ STATIC_INLINE char *jl_symbol_name_(jl_sym_t *s) JL_NOTSAFEPOINT
 
 static inline uint32_t jl_fielddesc_size(int8_t fielddesc_type) JL_NOTSAFEPOINT
 {
+    assert(fielddesc_type == 0 || fielddesc_type == 1 || fielddesc_type == 2);
     return 2 << fielddesc_type;
     //if (fielddesc_type == 0) {
     //    return sizeof(jl_fielddesc8_t);
@@ -1061,6 +1062,7 @@ static inline const char *jl_dt_layout_ptrs(const jl_datatype_layout_t *l) JL_NO
             return ((const jl_fielddesc16_t*)jl_dt_layout_fields(ly))[i].f;   \
         }                                                                     \
         else {                                                                \
+            assert(ly->fielddesc_type == 2);                                  \
             return ((const jl_fielddesc32_t*)jl_dt_layout_fields(ly))[i].f;   \
         }                                                                     \
     }                                                                         \
```
Note that I believe these assertions could be triggered already today: namely, when using a foreign type in the top level, right now precompilation will happy "serialize" it, and then loading the result of precompilation causes a segfault (this is easy to trigger with `GAP.jl`, which AFAIK is the only Julia package using foreign types so far). With the above assertion, at least in a debug build this would be caught earlier. I really hope that we can eventually submit another PR at some point which deals better with this: first approximation should be to error out; better would be if we can come up with a way to actually implementation (de)serialization for foreign types (and their instances), possibly via some more hooks (yes yes, in general serializing arbitrary GAP objects would be difficult, but at least for the basic ones that'd be doable). Anyway, this is a longterm thing; on the short term, though, I hope we can at least add some graceful error checks?

**Comment by fingolfin** [comment]
@rbehrends submitted a fix in PR #38510
